object-construction-checker/src/main/java/org/checkerframework/checker/objectconstruction/MustCallInvokedChecker.java

```diff
@@ -543,12 +585,148 @@ private void handleAssignFromRHS(
+            : cmValue.getAnnotations().stream()
+                .filter(anno -> AnnotationUtils.areSameByClass(anno, CalledMethods.class))
+                .findAny()
+                .orElse(typeFactory.top);
+
+    if (!calledMethodsSatisfyMustCall(mcValues, cmAnno)) {
+      Element lhsElement = TreeUtils.elementFromTree(lhs.getTree());
+      checker.reportError(
+          node.getTree(),
+          "required.method.not.called",
+          formatMissingMustCallMethods(mcValues),
+          lhsElement.asType().toString(),
+          " Non-final owning field might be overwritten");
+    }
+  }
+
+  /**
+   * Checks that the method that encloses an assignment is marked with @ResetMustCall annotation
+   * whose target is the object whose field is being re-assigned.
+   *
+   * @param node an assignment node whose lhs is a non-final, owning field
+   */
+  private void checkEnclosingMethodIsResetMC(AssignmentNode node) {
+    Node lhs = node.getTarget();
+    if (!(lhs instanceof FieldAccessNode)) {
+      return;
+    }
+
+    String receiverString = receiverAsString((FieldAccessNode) lhs);
+
+    TreePath currentPath = typeFactory.getPath(node.getTree());
+    MethodTree containingMethod = TreePathUtil.enclosingMethod(currentPath);
+    if (containingMethod == null) {
+      // Assignments outside of methods must be in constructors or field initializers, which
+      // are always safe.
+      return;
+    }
+    ExecutableElement enclosingMethod = TreeUtils.elementFromDeclaration(containingMethod);
+    AnnotationMirror resetMustCall =
+        typeFactory.getDeclAnnotation(enclosingMethod, ResetMustCall.class);
+    if (resetMustCall == null) {
+      checker.reportError(
+          containingMethod,
+          "missing.reset.mustcall",
+          receiverString,
+          ((FieldAccessNode) lhs).getFieldName());
+      return;
+    }
+
+    String targetStrWithoutAdaptation =
+        AnnotationUtils.getElementValue(resetMustCall, "value", String.class, true);
+    JavaExpressionContext context =
+        JavaExpressionParseUtil.JavaExpressionContext.buildContextForMethodDeclaration(
+            containingMethod, currentPath, checker);
+    String targetStr =
+        MustCallTransfer.standardizeAndViewpointAdapt(
+            targetStrWithoutAdaptation, currentPath, context);
+    if (!targetStr.equals(receiverString)) {
+      checker.reportError(
+          containingMethod,
+          "incompatible.reset.mustcall",
+          receiverString,
+          ((FieldAccessNode) lhs).getFieldName(),
+          targetStr);
+    }
+  }
+
+  /** Gets a standardized name for an object whose field is being re-assigned. */
+  private String receiverAsString(FieldAccessNode lhs) {
+    Node receiver = lhs.getReceiver();
+    if (receiver instanceof ThisNode) {
+      return "this";
+    }
+    if (receiver instanceof LocalVariableNode) {
+
+      return ((LocalVariableNode) receiver).getName();
+    }
+    throw new BugInCF(
```

Comment:
Should we really be crashing in this case?  Can't we just report an `incompatible.reset.mustcall` error?

Reply:
I originally wrote this as a crash to try to figure out what else it might be, but it was never anything else besides "this" or a LocalVariable in our tests.

I think I'd prefer to leave it as a crash, since if we ever encounter that crash we'll be able to modify this method to handle that kind of receiver - that's better than silently failing, IMO.

Reply:
Sure, that sounds good for now